Doc/whatsnew/3.13.rst

```diff
@@ -484,45 +484,137 @@ Optimizations
+IR:
+
+* Type propagation -- through forward data-flow analysis, we infer
+  and deduce information about types. This allows us to eliminate
+  much of the overhead associated with dynamic typing in the future.
+
+* Constant propagation -- through forward data-flow analysis, we can reduce
+  expressions like ::
 
-The internal architecture is roughly as follows.
+    a = 1
+    b = 2
+    c = a + b
 
-* We start with specialized *Tier 1 bytecode*.
-  See :ref:`What's new in 3.11 <whatsnew311-pep659>` for details.
+ to ::
 
-* When the Tier 1 bytecode gets hot enough, it gets translated
-  to a new, purely internal *Tier 2 IR*, a.k.a. micro-ops ("uops").
+    a = 1
+    b = 2
+    c = 3
 
-* The Tier 2 IR uses the same stack-based VM as Tier 1, but the
-  instruction format is better suited to translation to machine code.
+* Guard elimination -- through a combination of constant and type information,
+  we can eliminate type checks and other guards associated with operations.
+  These guards validate specialized operations, but add a slight bit of
+  overhead. For example, integer addition needs a type check that checks
+  both operands are integers. As a proof of concept, we managed to eliminate
+  over 70% of integer type checks in our own benchmarks.
 
-* We have several optimization passes for Tier 2 IR, which are applied
-  before it is interpreted or translated to machine code.
+* Loop splitting -- after the first iteration, we gain a lot more type
+  information. Thus, we peel the first iteration of loops to produce
+  an optimized body that exploits this additional type information.
+  This also achieves a similar effect to an optimization called
+  loop-invariant code motion, but only for guards.
 
-* There is a Tier 2 interpreter, but it is mostly intended for debugging
-  the earlier stages of the optimization pipeline. If the JIT is not
-  enabled, the Tier 2 interpreter can be invoked by passing Python the
-  ``-X uops`` option or by setting the ``PYTHON_UOPS`` environment
-  variable to ``1``.
+* Globals to constant promotion -- global value loads become constant
+  loads, speeding them up and also allowing for more constant propagation.
 
-* When the ``--enable-experimental-jit`` option is used, the optimized
-  Tier 2 IR is translated to machine code, which is then executed.
-  This does not require additional runtime options.
+* This section is non-exhaustive and will be updated with further
+  optimizations, up till CPython 3.13's beta release.
 
-* The machine code translation process uses an architecture called
-  *copy-and-patch*. It has no runtime dependencies, but there is a new
-  build-time dependency on LLVM.
+(Tier 2 optimizer contributed by Ken Jin, with implementation help
+by Guido van Rossum, Mark Shannon, and Jules Poon. Special thanks
+to Manuel Rigger and Martin Henz.)
+
+
+Execution Engine
+----------------
+
+There are two execution engines for Tier 2 IR.
```

Comment:
```suggestion
There are two execution engines for Tier 2 IR: the Tier 2 interpreter and the Just-in-Time (JIT) compiler.
```

It's jarring to see the JIT mentioned in the paragraph below before it is introduced, so I think introducing them both here helps.